lib/spack/spack/util/executable.py

```diff
@@ -269,39 +270,38 @@ def __str__(self):
         return " ".join(self.exe)
 
 
-@system_path_filter
 def which_string(*args, **kwargs):
     """Like ``which()``, but return a string instead of an ``Executable``."""
     path = kwargs.get("path", os.environ.get("PATH", ""))
     required = kwargs.get("required", False)
 
-    if isinstance(path, str):
-        path = path.split(os.pathsep)
+    if isinstance(path, list):
+        paths = [Path(str(x)) for x in path]
 
-    for name in args:
-        win_candidates = []
-        if sys.platform == "win32" and (not name.endswith(".exe") and not name.endswith(".bat")):
-            win_candidates = [name + ext for ext in [".exe", ".bat"]]
-        candidate_names = [name] if not win_candidates else win_candidates
+    if isinstance(path, str):
+        paths = [Path(x) for x in path.split(os.pathsep)]
 
+    for search_item in args:
+        search_item = Path(search_item)
+        candidate_items = [Path(search_item)]
         if sys.platform == "win32":
-            new_path = path[:]
-            for p in path:
-                if os.path.basename(p) == "bin":
-                    new_path.append(os.path.dirname(p))
-            path = new_path
-
-        for candidate_name in candidate_names:
-            if os.path.sep in candidate_name:
-                exe = os.path.abspath(candidate_name)
-                if os.path.isfile(exe) and os.access(exe, os.X_OK):
-                    return exe
-            else:
-                for directory in path:
-                    directory = path_to_os_path(directory).pop()
-                    exe = os.path.join(directory, candidate_name)
-                    if os.path.isfile(exe) and os.access(exe, os.X_OK):
-                        return exe
+            if not search_item.suffix:
+                candidate_items = [
+                    search_item.parent / (search_item.name + ext) for ext in [".exe", ".bat"]
+                ]
+
+            paths_copy = paths
+            for p in paths_copy:
+                if p.name == "bin":
+                    paths.append(p.parent)
+
+        for candidate_item in candidate_items:
+            search_paths = [Path.cwd()] + paths
```

Comment:
You mention

> Note: to make behavior similar for both windows and linux, this class now adds the current working directory to the path when finding the executable.

Does Windows search `.` by default for executables? According to https://superuser.com/questions/238987/how-does-unix-search-for-executable-files, this is not the case for Linux, and it doesn't seem to be the case for Mac OS either.

Reply:
On windows, the cwd is searched and then if the file isn't found, the path will be searched. Linux doesn't look in the cwd directly at all and only looks in the path. To look for something in the cwd on linux, you'd use `./my_exe`. The pathlib cast in the `__init__` function strips any preceding `./` which is why this behavior change is in place.

If this behavior change doesn't seem intuitive for Linux or Mac, we could only add the cwd when `./` is in the executable path. Another option @johnwparent had mentioned is adding the cwd to the end of path but this would lead to `./` executables being shadowed by something of the same name in the path.

Reply:
On second thought, my suggestions doesn't make a ton of sense. Disregard.

Reply:
There are a number of solutions that work here if we have access to the original string (such as detecting the dot slash and resolving the absolute path based on the CWD of the `__call__` method or only adding the CWD to the path if we see the dot) but I feel like we'll need a more generic solution once the rest of Spack proper is refactored to pathlib.

Reply:
I added changes to only use the CWD at `__init__` time and resolve the absolute path for the executable.